subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_os_grp
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+#include <pm_config.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt.h>
+#include <zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h>
+
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+
+
+static int os_reset_status;
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static int reset_res_fn(struct net_buf *nb, void *user_data)
+{
+	if (!nb) {
+		os_reset_status = MGMT_ERR_ETIMEOUT;
+	} else {
+		os_reset_status = MGMT_ERR_EOK;
+	}
+	k_sem_give(user_data);
+	return 0;
+}
+
+int os_mgmt_client_reset(void)
+{
+	struct net_buf *nb;
+	int rc;
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	/* allocate buffer */
+	nb = smp_client_buf_allocation(MGMT_GROUP_ID_OS, OS_MGMT_ID_RESET, MGMT_OP_WRITE);
+	if (!nb) {
+		rc = MGMT_ERR_ENOMEM;
+		goto end;
+	}
+	os_reset_status = -1;
+	k_sem_reset(&spm_mutex);
+	rc = smp_client_send_cmd(nb, reset_res_fn, &spm_mutex, CONFIG_SMP_CMD_DEFAULT_LIFE_TIME);
+	if (rc) {
+		smp_packet_free(nb);
+		goto end;
+	}
+	/* Wait for process end update event */
+	k_sem_take(&spm_mutex, K_FOREVER);
+	rc = os_reset_status;
+end:
+	k_mutex_unlock(&mcumgr_mutex);
+	return rc;
+}
+
+#ifdef CONFIG_MCUMGR_GRP_OS_ECHO_REQ
+
+static const char *echo_cmp_string;
+static int echo_status;
+/* Mimimal number of ZCBOR encoder states to provide full encoder functionality plus one for backup
+ */
+#define CBOR_ENCODER_OS_STATE_NUM 3
```

Comment:
put all these at the top of the file with other variable/define declarations (copy the #ifdef up there)

Reply:
This file is only builded when CONFIG_MCUMGR_CLIENT_GRP_OS is defined. so up of the src file not need #ifdef CONFIG_MCUMGR_GRP_OS_ECHO_REQ